lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,128 @@
+    USER_APP_VERSION = 'USER_APP_VERSION'
+
+    def process_request(self, request):
+        """
+        raises HTTP Upgrade Require error if request is from mobile native app and
+        user app version is no longer supported
+        """
+        version_data = self.get_info_for_version(request)
+        if version_data:
+            last_supported_date = version_data.get(self.UPGRADE_DEADLINE_HEADER, None)
+            if last_supported_date and last_supported_date != self.NO_LAST_SUPPORTED_DATE:
+                if datetime.now().replace(tzinfo=UTC) > last_supported_date:
+                    return HttpResponse(status=426)
+
+    def process_response(self, _request, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < expiry of that version
+        """
+        version_data = self.get_info_for_version(_request)
+        if version_data:
+            upgrade_deadline = version_data.get(self.UPGRADE_DEADLINE_HEADER, None)
+            if upgrade_deadline and upgrade_deadline != self.NO_LAST_SUPPORTED_DATE:
+                response[self.UPGRADE_DEADLINE_HEADER] = upgrade_deadline
+            latest_version = version_data.get(self.LATEST_VERSION_HEADER, None)
+            user_app_version = version_data.get(self.USER_APP_VERSION, None)
+            if (latest_version and latest_version != self.NO_LATEST_VERSION and
+                    parsed_version(user_app_version) < parsed_version(latest_version)):
+                response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
+
+    def get_cache_key_name(self, user_agent, field):
+        """ get key name to use to cache any property against user agent """
+        return "mobile_api.app_version_upgrade.{}.{}".format(field, user_agent)
+
+    def get_info_for_version(self, request):
+        """
+        It sets request cache data for last_supported_date and latest_version with memcached values if exists against
+        user-agent else computes the values for specific platform and sets it in both memcache (for next
+        server interaction from same user-agent/platform) and request cache
+        """
+        request_cache_dict = RequestCache.get_request_cache().data
+        last_supported_date = request_cache_dict.get(self.UPGRADE_DEADLINE_HEADER, None)
+        latest_version = request_cache_dict.get(self.LATEST_VERSION_HEADER, None)
+        user_app_version = request_cache_dict.get(self.USER_APP_VERSION, None)
+        if last_supported_date and latest_version and user_app_version:
+            return {
+                self.UPGRADE_DEADLINE_HEADER: last_supported_date,
+                self.LATEST_VERSION_HEADER: latest_version,
+                self.USER_APP_VERSION: user_app_version,
+            }
```

Comment:
Instead of saving the data in the root of the request cache, it will be simpler to store it in a dict and save that in the RequestCache against "mobile_api.app_version_upgrade.info_for_version". Then you just need to check if that key exists in the `request_cache` and if it does return it. Similarly, at the end the dict just needs to be put into the `request_cache` and then returned.